pkg/controller/openapi/openapi_controller.go

```diff
@@ -0,0 +1,399 @@
+		return statusReconciler, reconcile.Result{}, err
+	}
+
+	statusReconciler := NewOpenAPIStatusReconciler(r.BaseReconciler, openapiCR, providerAccount.AdminURLStr, err, productSynced)
+	return statusReconciler, reconcile.Result{Requeue: !productSynced}, err
+}
+
+func (r *ReconcileOpenAPI) validateSpec(resource *capabilitiesv1beta1.OpenAPI) error {
+	errors := field.ErrorList{}
+	errors = append(errors, resource.Validate()...)
+
+	if len(errors) == 0 {
+		return nil
+	}
+
+	return &helper.SpecFieldError{
+		ErrorType:      helper.InvalidError,
+		FieldErrorList: errors,
+	}
+}
+
+func (r *ReconcileOpenAPI) checkProductSynced(resource *capabilitiesv1beta1.OpenAPI) (bool, error) {
+	if resource.Status.ProductResourceName == nil {
+		// product resource name not available to check
+		return false, nil
+	}
+
+	// Fetch the Product instance
+	product := &capabilitiesv1beta1.Product{}
+	objectKey := client.ObjectKey{Name: resource.Status.ProductResourceName.Name, Namespace: resource.Namespace}
+	err := r.Client().Get(r.Context(), objectKey, product)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		// Error reading the object - requeue the request.
+		return false, err
+	}
+
+	return product.Status.Conditions.IsTrueFor(capabilitiesv1beta1.ProductSyncedConditionType), nil
+}
+
+func (r *ReconcileOpenAPI) readOpenAPI(resource *capabilitiesv1beta1.OpenAPI) (*openapi3.Swagger, error) {
+	// OpenAPIRef is oneOf by CRD openapiV3 validation
+	if resource.Spec.OpenAPIRef.ConfigMapRef != nil {
+		return r.readOpenAPIConfigMap(resource)
+	}
+
+	// Must be URL
+	return r.readOpenAPIFromURL(resource)
+}
+
+func (r *ReconcileOpenAPI) readOpenAPIConfigMap(resource *capabilitiesv1beta1.OpenAPI) (*openapi3.Swagger, error) {
+	fieldErrors := field.ErrorList{}
+	specFldPath := field.NewPath("spec")
+	openapiRefFldPath := specFldPath.Child("openapiRef")
+	configMapRefFldPath := openapiRefFldPath.Child("configMapRef")
+
+	objectKey := client.ObjectKey{Namespace: resource.Spec.OpenAPIRef.ConfigMapRef.Namespace, Name: resource.Spec.OpenAPIRef.ConfigMapRef.Name}
+	openapiConfigMapObj := &corev1.ConfigMap{}
+
+	// Read config map
+	if err := r.Client().Get(r.Context(), objectKey, openapiConfigMapObj); err != nil {
+		if errors.IsNotFound(err) {
+			fieldErrors = append(fieldErrors, field.Invalid(configMapRefFldPath, resource.Spec.OpenAPIRef.ConfigMapRef, "ConfigMap not found"))
+			return nil, &helper.SpecFieldError{
+				ErrorType:      helper.InvalidError,
+				FieldErrorList: fieldErrors,
+			}
+		}
+
+		// unexpected error
+		return nil, err
+	}
+
+	if len(openapiConfigMapObj.Data) < 1 {
+		fieldErrors = append(fieldErrors, field.Invalid(configMapRefFldPath, resource.Spec.OpenAPIRef.ConfigMapRef, "ConfigMap was empty"))
+		return nil, &helper.SpecFieldError{
+			ErrorType:      helper.InvalidError,
+			FieldErrorList: fieldErrors,
+		}
+	}
+
+	// Get arbitrary key value
```

Comment:
Is the intended approach to get the first attribute found in the ConfigMap, independently of the attribute name?

Reply:
Not the first one, but an arbitrary key. 

It is explained in the doc and reference doc. If the secret contains several fields (files), only one is read.

Reply:
I mean the first one processed by the code. `Data` is a map, there's no order, so basically it will randomly read one entry among the entries if there is more than one of them.

Reply:
Check doc. The implementation is consistent with the doc. Is the doc incorrect? Should the secret be processed otherwise?

Reply:
I checked the doc already on my first reply.

My original comment was asking confirmation about the implementation and based on the answers I think my explanation is accurate.

If we are talking about the doc, it's not wrong, it's not very accurate.

```
The secret must have at least one field with value set to the openapi document content. The field name will not be read.
```

If it has more than one field, the field to be read will be implementation-specific (random order). Maybe it would be safer to document that it has to have `one field`, and not `at least one field`. I'm fine with current content though so up to you. Again, the original comment was not about doc, was about asking for confirmation of understanding about the implementation.

Reply:
Added doc and required validation so the secret only contains a single field

Reply:
:+1: